File: sol/solution/0101-0200/0106/Solution.php
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     public $val = null;
 *     public $left = null;
 *     public $right = null;
 *     function __construct($val = 0, $left = null, $right = null) {
 *         $this->val = $val;
 *         $this->left = $left;
 *         $this->right = $right;
 *     }
 * }
 */
class Solution {
    private $postIndex;
    private $inMap;
    /**
     * @param Integer[] $inorder
     * @param Integer[] $postorder
     * @return TreeNode
     */
    function buildTree($inorder, $postorder) {
        $this->inMap = [];
        foreach ($inorder as $i => $val) {
            $this->inMap[$val] = $i;
        }
        $this->postIndex = count($postorder) - 1;
        return $this->buildTreeHelper($inorder, $postorder, 0, count($inorder) - 1);
    }
    private function buildTreeHelper($inorder, $postorder, $inStart, $inEnd) {
        if ($inStart > $inEnd) return null;
        $root = new TreeNode($postorder[$this->postIndex--]);
        $inIndex = $this->inMap[$root->val];
        $root->right = $this->buildTreeHelper($inorder, $postorder, $inIndex + 1, $inEnd);
        $root->left = $this->buildTreeHelper($inorder, $postorder, $inStart, $inIndex - 1);
        return $root;
    }
}

File: sol/solution/0101-0200/0106/Solution.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.left = nil
 *         self.right = nil
 *     }
 * }
 */
class Solution {
    private var postIndex: Int = 0
    private var inMap: [Int: Int] = [:]
    func buildTree(_ inorder: [Int], _ postorder: [Int]) -> TreeNode? {
        for (i, val) in inorder.enumerated() {
            inMap[val] = i
        }
        postIndex = postorder.count - 1
        return buildTreeHelper(inorder, postorder, 0, inorder.count - 1)
    }
    private func buildTreeHelper(_ inorder: [Int], _ postorder: [Int], _ inStart: Int, _ inEnd: Int) -> TreeNode? {
        if inStart > inEnd {
            return nil
        }
        let root = TreeNode(postorder[postIndex])
        postIndex -= 1
        let inIndex = inMap[root.val]!
        root.right = buildTreeHelper(inorder, postorder, inIndex + 1, inEnd)
        root.left = buildTreeHelper(inorder, postorder, inStart, inIndex - 1)
        return root
    }
}